### PR TITLE
Sign in

### PR DESCRIPTION
**Pull request by jxczz2**
## Summary

+ signIn is added
+ validations signIn is added
+ Test signIn is added 

## Known Issues

No problem.

## Trello Card

https://trello.com/c/xihe253w

### CODE REVIEW

**Review by picollomartin — CHANGES_REQUESTED** [review]
Same as before, compare with original branch not with master

File: app/controllers/users.js [hunk 1]
```diff
@@ -10,11 +14,46 @@ exports.userCreate = (req, res, next) => {
...
+  return User.createUser(user)
     .then(() => {
       res.status(201).end();
     })
     .catch(err => {
       next(err);
     });
 };
+
+const giveToken = req => {
```
Comment (picollomartin) [inline app/controllers/users.js]: Make logic for tokens as a service

File: app/controllers/users.js [hunk 2]
```diff
@@ -10,11 +14,46 @@ exports.userCreate = (req, res, next) => {
...
+    password: req.body.password
+  };
+  try {
+    const result = await User.getUserBy(user.email);
+    if (!result) throw error.signInError('user not registered');
+    return bcrypt.compare(user.password, result.password, (err, validPassword) => {
+      if (validPassword) {
+        const token = giveToken(user);
+        res
+          .cookie('accesToken', token)
```
Comment (picollomartin) [inline app/controllers/users.js]: Use a header for return token, `authorization` is normally used (make this a config variable)

File: app/controllers/users.js [hunk 3]
```diff
@@ -10,11 +14,46 @@ exports.userCreate = (req, res, next) => {
...
+          .send(`x-access-token ${token}`)
```
Comment (picollomartin) [inline app/controllers/users.js]: This is unnecessary

File: app/models/user.js [hunk 4]
```diff
@@ -36,5 +36,16 @@ module.exports = (sequelize, DataTypes) => {
         throw errors.databaseError(err);
       });
   };
+
+  User.getUserBy = email =>
+    User.findOne({
+      where: {
+        email
```
Comment (picollomartin) [inline app/models/user.js]: Make a generic condition here

File: app/routes.js [hunk 5]
```diff
@@ -1,6 +1,8 @@
 const signUpMiddleware = require('./middlewares/signUpValidations'),
+  signInMiddleware = require('./middlewares/signInValidations'),
   userController = require('./controllers/users');
 
 exports.init = app => {
-  app.post('/users', signUpMiddleware.signUpValidate, userController.userCreate);
+  app.post('/users', signUpMiddleware.checkValidations, userController.userCreate);
+  app.post('/users/sessions/', signInMiddleware.checkValidations, userController.sesion);
```
Comment (picollomartin) [inline app/routes.js]: Remove trailing `/`

Comment (picollomartin) [reply app/routes.js]: Check typo in sesion

File: app/middlewares/signInValidations.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+const { validateUser } = require('./validations'),
+  error = require('../errors');
+
+exports.checkValidations = (req, res, next) => {
```
Comment (picollomartin) [inline app/middlewares/signInValidations.js]: Do you need an entire file for this ?, also we talk about a generic middleware for validators (not a function for each validation)

**Review by picollomartin — CHANGES_REQUESTED** [review]
Check little typos in code, like acces or sesion (if you want, install a [spell checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) for vs-code)

File: app/controllers/users.js (hunk 3 above)
Comment (jxczz2) [reply app/controllers/users.js]: remove send ?

File: app/controllers/users.js [hunk 7]
```diff
@@ -10,11 +15,44 @@ exports.userCreate = (req, res, next) => {
...
     });
 };
+
+exports.session = async (req, res, next) => {
+  const user = {
+    email: req.body.email,
+    password: req.body.password
+  };
+  try {
+    const signErrors = validateUser(user, ['email', 'password']);
```
Comment (picollomartin) [inline app/controllers/users.js]: Why this validation is here ?

Comment (picollomartin) [reply app/controllers/users.js]: This validation should be in the middleware

**Review by picollomartin — CHANGES_REQUESTED** [review]

File: app/routes.js [hunk 8]
```diff
@@ -1,6 +1,11 @@
-const signUpMiddleware = require('./middlewares/signUpValidations'),
+const {
+    validatorMiddleware,
+    signUpCheckValidations,
+    signInCheckValidations
+  } = require('./middlewares/signValidations'),
```
Comment (picollomartin) [inline app/routes.js]: Rename this middleware

**Review by picollomartin — APPROVED** [review]
Check some name of variables/functions, things like `result` or `session` (for a controller function) sometimes are a little tricky to understand

**Pull request merged**